tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/TensorScope.java

```diff
@@ -0,0 +1,211 @@
+/*
+  Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ==============================================================================
+ */
+package org.tensorflow;
+
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.Set;
+import java.util.WeakHashMap;
+
+
+/**
+ * A scope that can be used to manage tensor resources.  If auto-attach is used, any tensors created between a scope's
+ * creation and calling {@code close()}, that haven't been detached, are guaranteed to be closed with the scope (even if
+ * they are created in a sub-scope).  Tensors may be manually closed earlier without issue.
+ * <p>
+ * When auto-attach is true, tensors are automatically tracked on creation.  A tensor can me manually added to a scope
+ * with {@link TensorScope#attach(Tensor)} or {@link Tensor#attachToCurrentScope()}.  The tensor will then be closed when the first of it's managing scopes closes.
+ * <p>
+ * {@link Tensor#detach()} detaches the tensor from all scopes, requiring the user to close it manually or attach it to
+ * another scope.
+ * <p>
+ * Note that scope management is thread local, except for detach, which will detach even from scopes on other threads.
```

Comment:
Why do we want the thread local scoping? How is it going to interact with Loom?

Reply:
Well, Loom is supposed to give us an alternative to TLS:
http://cr.openjdk.java.net/~rpressler/loom/loom/sol1_part2.html#scope-variables
But that's not going to happen anytime soon, so in the mean time we need *something*... It sounds to me though that if we were to replace TLS with something like that, we would need to juggle 3 types of objects: `Tensor`, `TensorScope`, and something like `TensorScopeScope`... Am I reading this correctly?

Reply:
`Scoped` would be perfect for this, I think.  Probably rather than a `TensorScopeScope` I'd have it add itself to the current scope as a child on creation, or something like that.

Reply:
Well I assume it's thread local to avoid a bunch of synchronisation, but it's not clear why we'd want the scopes to be thread local. What's the argument for that behaviour?

Reply:
Because it's a lexical scope (like `{}`), representing a block of code or certain segment of execution.  It doesn't make any sense to do something like
```java
try(TensorScope scope = new TensorScope()){
  myTensorFunction();
}
```
and have it pick up tensors created by something else going on in another thread.

My latest changes should alleviate the issue somewhat, I'm using `InheritableThreadLocal` for the current scope, but the hierarchy can cross threads, so if you create new threads in a scope like above, they will add tensors to the scope, but previously existing threads don't.